lib/pbio/drv/usb/usb_stm32.c

```diff
@@ -139,13 +295,75 @@ pbdrv_usb_bcd_t pbdrv_usb_get_bcd(void) {
     return pbdrv_usb_bcd;
 }
 
+static void pbdrv_usb_serial_transmit(void) {
+    static uint32_t tx_size = 0;
+
+    if (usb_out_busy) {
+        return;
+    }
+
+    // If tx_size > 0 it means we have a pending retry, otherwise we get as
+    // much as we can from the stdout buffer.
+    if (tx_size == 0) {
+        tx_size = ringbuf_elements(&stdout_buf);
+        if (tx_size > PBIO_ARRAY_SIZE(usb_out_buf)) {
+            tx_size = PBIO_ARRAY_SIZE(usb_out_buf);
+        }
+        if (tx_size > 0) {
+            for (uint32_t i = 0; i < tx_size; i++) {
+                usb_out_buf[i] = ringbuf_get(&stdout_buf);
+            }
+        }
+    }
+
+    if (tx_size > 0) {
+        USBD_CDC_SetTxBuffer(&husbd, usb_out_buf, tx_size);
+        if (USBD_CDC_TransmitPacket(&husbd) == USBD_OK) {
+            usb_out_busy = true;
+            tx_size = 0;
+        }
+    }
+}
+
+static void pbdrv_usb_serial_receive(void) {
+    if (usb_in_busy) {
+        return;
+    }
+
+    if (USBD_CDC_ReceivePacket(&husbd) == USBD_OK) {
+        usb_in_busy = true;
+    }
+}
+
+// TODO: need to multiplex stdin/stdout with Bluetooth or remove USB serial
+
+pbio_error_t pbsys_usb_put_char(uint8_t c) {
+    if (!usb_connected) {
+        // don't lock up print() when USB not connected - data is discarded
+        return PBIO_SUCCESS;
+    }
+    if (ringbuf_put(&stdout_buf, c) == 0) {
+        return PBIO_SUCCESS;
+    }
+    return PBIO_SUCCESS;
+}
+
+pbio_error_t pbsys_usb_get_char(uint8_t *c) {
+    if (ringbuf_elements(&stdin_buf) == 0) {
+        return PBIO_ERROR_AGAIN;
+    }
+    *c = ringbuf_get(&stdin_buf);
+    return PBIO_SUCCESS;
+}
+
 // Event loop
 
 PROCESS_THREAD(pbdrv_usb_process, ev, data) {
     static struct pt bcd_pt;
     static PBIO_ONESHOT(no_vbus_oneshot);
     static PBIO_ONESHOT(bcd_oneshot);
     static bool bcd_busy;
+    static struct etimer timer;
```

Comment:
It looks like the call to `etimer_set()` is missing. But it would be event better if this was interrupt driven rather than polled with a timer.

Reply:
Using an interrupt is beyond my level of understanding at this point, but if you have an example of how this would work I'm willing to look into it!

Reply:
In the template I linked above, there are functions like `USBD_TEMPLATE_DataIn` that says `handle ...` in the doc comment. I haven't dug into the code but I'm pretty sure these are called in an interrupt context. So they need to do the absolute minimum, like copy data received to the ring buffer or set a voltatile flag, then call `process_poll()` to notify the contiki process that there is now new info to be processed.